### PR TITLE
Support allreduce fusion with cp

### PR DESCRIPTION
**Pull request by Shunkangz**
## Motivation

Enable the allreduce_fusion with context parallel. This requires the allreduce_fusion can work with any sub comm group.

## Modifications

This PR can only work after flashInfer [PR](https://github.com/flashinfer-ai/flashinfer/pull/2662) merged. 

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/communicator.py [hunk 1]
```diff
@@ -460,7 +457,7 @@ def prepare_attn(
                 ) and hasattr(self.input_layernorm, "forward_with_allreduce_fusion"):
                     hidden_states, residual = (
                         self.input_layernorm.forward_with_allreduce_fusion(
-                            hidden_states, residual, use_attn_tp_group=True
+                            hidden_states, residual, use_attn_tp_group=False
```
Comment (Fridge003) [inline python/sglang/srt/layers/communicator.py:518]: Why changing it to false here?

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -2098,7 +2098,6 @@ def _handle_model_specific_adjustments(self):
             and (is_sm90_supported() or is_sm100_supported())
             and self.tp_size > 1
             and not self.enable_dp_attention
-            and self.attn_cp_size <= 1
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py:2217]: Is this case covered by `test_qwen3_30b.py`? BTW I feel it's better to rename this test as `test_cp_small.py`

File: python/sglang/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -2035,6 +2035,27 @@ def _warmup_fused_sampling(self):
             logits_dtype=logits_warmup_dtype,
         )
 
+    def _pre_initialize_flashinfer_allreduce_workspace(self):
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/model_runner.py:2185]: This function looks like a dead code not called anywhere else

File: python/sglang/srt/layers/flashinfer_comm_fusion.py [hunk 4]
```diff
@@ -211,12 +267,51 @@ def cleanup(self):
...
                 self.world_size = None
                 self.rank = None
+                self.group = None
                 self.max_token_num = None
                 self.hidden_dim = None
                 self.dtype = None
 
 
-_workspace_manager = FlashInferWorkspaceManager()
+_attn_tp_workspace_manager = FlashInferWorkspaceManager()
```
Comment (Fridge003) [inline python/sglang/srt/layers/flashinfer_comm_fusion.py:276]: Can we create these managers lazily. Since we don't need both of them at the same time

File: python/sglang/srt/layers/flashinfer_comm_fusion.py [hunk 5]
```diff
@@ -385,7 +502,46 @@ def flashinfer_allreduce_residual_rmsnorm(
     return norm_out, residual_out
 
 
+def pre_initialize_workspaces(
```
Comment (Fridge003) [inline python/sglang/srt/layers/flashinfer_comm_fusion.py:505]: This also looks like dead code

File: python/sglang/srt/layers/flashinfer_comm_fusion.py [hunk 6]
```diff
@@ -234,46 +329,67 @@ def ensure_workspace_initialized(
...
             world_size = get_moe_tensor_parallel_world_size()
             rank = get_moe_tensor_parallel_rank()
+            coordinator = get_moe_tp_group()
+
+    # When the sub-group IS the full TP group, pass None so the workspace
+    # uses the default process group directly (no TorchDistBackend needed).
+    # For true sub-groups, use NCCL device_group for GPU/device mapping and
+    # GLOO cpu_group for metadata broadcasts (avoids NCCL collectives that
+    # interfere with CUDA graph capture).
+    if coordinator.device_group is tp_coordinator.device_group:
```
Comment (Fridge003) [inline python/sglang/srt/layers/flashinfer_comm_fusion.py:353]: Looks like the creation of coordinator/device_gropup/cpu_group can only be executed once, so we had better put them under the if branch at line 367

File: python/sglang/srt/layers/flashinfer_comm_fusion.py [hunk 7]
```diff
@@ -385,7 +502,46 @@ def flashinfer_allreduce_residual_rmsnorm(
...
+
+    This must be called before graph capture to avoid collective operations
+    (broadcasts, barriers) inside the graph capture context, which can
+    deadlock with custom_all_reduce.register_graph_buffers.
+    """
+    if _flashinfer_allreduce_unavailable or _flashinfer_comm is None:
+        return
+
+    # Initialize MoE workspace
+    ensure_workspace_initialized(
```
Comment (Fridge003) [inline python/sglang/srt/layers/flashinfer_comm_fusion.py:521]: Why ensure twice

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/communicator.py (hunk 1 above)
Comment (Shunkangz) [reply python/sglang/srt/layers/communicator.py:518]: It is a previous bug. This is pre attention allreduce and we should do it cross MoE group instead of Attention Group.

File: python/sglang/srt/server_args.py (hunk 2 above)
Comment (Shunkangz) [reply python/sglang/srt/server_args.py:2217]: Yes. It is enabled by default. I can rename it.

File: python/sglang/srt/model_executor/model_runner.py (hunk 3 above)
Comment (Shunkangz) [reply python/sglang/srt/model_executor/model_runner.py:2185]: Fix it.

File: python/sglang/srt/layers/flashinfer_comm_fusion.py (hunk 4 above)
Comment (Shunkangz) [reply python/sglang/srt/layers/flashinfer_comm_fusion.py:276]: I follow the existing logic to initialize attn_tp work space eagerly. As for moe_tp workspace, I will initialize it when we need.

File: python/sglang/srt/layers/flashinfer_comm_fusion.py (hunk 5 above)
Comment (Shunkangz) [reply python/sglang/srt/layers/flashinfer_comm_fusion.py:505]: This is used in `model_runner.py`

File: python/sglang/srt/layers/flashinfer_comm_fusion.py (hunk 7 above)
Comment (Shunkangz) [reply python/sglang/srt/layers/flashinfer_comm_fusion.py:521]: One for Attention TP workspace and One for MoE TP workspace.

File: python/sglang/srt/layers/flashinfer_comm_fusion.py (hunk 6 above)
Comment (Shunkangz) [reply python/sglang/srt/layers/flashinfer_comm_fusion.py:353]: Because the check condition of the if branch depends on those metadata `coordinator/device_gropup/cpu_group ` So we should leave them here.

File: python/sglang/srt/layers/flashinfer_comm_fusion.py (hunk 7 above)
Comment (Fridge003) [reply python/sglang/srt/layers/flashinfer_comm_fusion.py:521]: We might skip the initialization here if any of attnetion TP/ MOE TP is not enabled

**Comment by Fridge003** [comment]
@Shunkangz Please check this failure https://github.com/sgl-project/sglang/actions/runs/24327366445/job/71025454116?pr=21249

File: python/sglang/srt/layers/flashinfer_comm_fusion.py (hunk 7 above)
Comment (Shunkangz) [reply python/sglang/srt/layers/flashinfer_comm_fusion.py:521]: Sure. The code has already covered this. If the target group size is smaller than 1, we will not allocate the workspace.

**Comment by Shunkangz** [comment]
> @Shunkangz Please check this failure https://github.com/sgl-project/sglang/actions/runs/24327366445/job/71025454116?pr=21249

It's a merge artifact. I fixed it.

**Review by Fridge003 — APPROVED** [review]
